Add internal binary reader/writer API for packet construction.

iphydf opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/toktok/c-toxcore/320)
<!-- Reviewable:end -->

nurupo commented:
Just the interface, no implementation?

---

Btw, don't think I ever got anything other than "too many macros" when asking you to review the `[writable|readable]_network_buffer` (`[w|r]nb`), which is very similar to this binary writer, except that it additionally performs byte-order conversion: from host to network on writing and from network to host when reading.

- [Final version](https://gist.github.com/nurupo/2228a603fb4b22a10d82fe5b1aa5ef1b)
  The final version of `[w|r]nb` uses a dynamically resizable array called `Vector`, which mimics C++'s std::vector interface but implements only a limited subset of it that is required for `[w|r]nb`. `Vector` was introduced so that you wouldn't have to specify the exact length of the buffer for `writable_network_buffer`, you could just write things into the buffer and it would resize the underlying array for you, e.g., using your interface from this PR, instead of doing `bin_w_new(sizeof(something) + 1 + 2 + 3 + 5 + 7 + 11 + (n * m) + sizeof(uint32_t) + sizeof(DHT_Node)*nodes);` you would just do `bin_w_new();` and you can write as many things as 2^16 bytes, which is the limit in my implementation that can be easily raised.
  - Usage of `writable_network_buffer`
    - create `wnb`
    - write things into it, it converts the byte order from host to network of the things you write
    - writing always appends data to the end
    - you can write as many thing as 2^16 bytes (that limit can be easily raised), the underlying array would automatically resize
    - when you are done writing, you can get detach (take over the memory ownership) the data written into it as a `Vector`
  - Usage of `readable_network_buffer`
    - create `rnb` by attaching (giving up memory ownership of) a `Vector` object with data in it
    - read things from it, it converts the byte order from network to host of the things you read
    - reading always takes data from the front
    - you can't read more than the `Vector` contains, you get an error returned if you try to
    - when you are done reading, you `free()` `rnb`, which frees the `Vector` you gave up the ownership of

- [Initial version](https://gist.github.com/nurupo/2228a603fb4b22a10d82fe5b1aa5ef1b/a4655a4e15f6e314cd7baeca02e674d9d1fbf538)
  This version doesn't use `Vector` and expects the user to allocate the memory on their own. This implementation can be used without any `free()` calls, i.e. you can allocate a VLA on stack, attach a `buffer_writer` onto the array, write things into the array using the `buffer_writer` and be done with it -- no need to deallocate anything, stack unwinding would destroy the `buffer_writer` and VLA. You need to know the length of the data you want to write into the array though, since you don't use `Vector`.
  - Usage of `buffer_writer`
    - create `bw` by passing it `(data, length)` pair of pre-allocated array
    - write things into it, it converts the byte order from host to network of the things you write
    - writing always appends data to the end
    - you can't write more than the `length` of the array, you get an error returned if you try to
    - when you are done writing -- you are done, as there is no cleanup step you need to perform
  - Usage of `buffer_reader`
    - create `br` by passing it `(data, length)` pair of array containing data in it
    - read things from it, it converts the byte order from network to host of the things you read
    - reading always takes data from the front
    - you can't read more than the `length` of the array, you get an error returned if you try to
    - when you are done reading -- you are done, as there is no cleanup step you need to perform

iphydf commented:
This intentionally does not have an implementation, because I want to discuss and develop the API, first. After that, we write tests, after that we implement it. I indeed don't agree with the many macros in your implementation, but we can discuss that in the third step. First we need an API that is useful for our use case. So, we can talk about the API and its documentation, and at the same time port some (most? all?) of toxcore's functions to it. The use case dictates the API.

iphydf commented:
Clarified the documentation about the byte order in which the writer serialises integers.

nurupo commented:
Review status: 0 of 4 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[toxcore/bin_io.h, line 55 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQWS-pDAP9FzBww9kY:-KZQWS-pDAP9FzBww9kZ:b-1lc17a) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L55)):*
> ```objc
> #ifndef BIN_W_DEFINED
> #define BIN_W_DEFINED
> typedef struct Bin_W Bin_W;
> ```

Why do you have `struct FOO foo` function arguments if you typedef'ed the struct?

---

*[toxcore/bin_io.h, line 66 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQacQkHsY67gfQIiy2:-KZQacQl-vNmEioaPNv-:b-assk4d) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L66)):*
> ```objc
>  * @param initial The initial capacity of the writer.
>  */
> struct Bin_W *bin_w_new(size_t initial, BIN_ERR_NEW *error);
> ```

How about having a constructor that doesn't require specification of the inital capacity?

Current constructor encourages to sum sizes of all things we are going to write, e.g. `bin_w_new(size1 + size2 + SOME_CONSTANT_SIZE*6 +1 +7 + size44 + sizeof(uint64_t), &error);`, which is quite ugly. If the writer automatically resizes internal array, this should really be optional. The default initial capacity could be the maximum toxcore packet size (there is a define for that) or something else if you think that's a bad choice.

---

*[toxcore/bin_io.h, line 79 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQbreQ96mp99VbbVzM:-KZQbreQ96mp99VbbVzN:b-f41qln) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L79)):*
> ```objc
>  * pointer becomes invalid.
>  */
> uint8_t *bin_w_get(const struct Bin_W *w);
> ```

Add: This pointer becomes invalid when `bin_w_u*()` or `bin_w_arr()` functions are called, as they might reallocate the internal buffer at a different address.

---

*[toxcore/bin_io.h, line 185 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQW_oIGfV_gzPspiFf:-KZQW_oIGfV_gzPspiFg:b34cfu6) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L185)):*
> ```objc
>  * Read an 8 bit unsigned integer and advance the reader by 1 byte.
>  */
> uint8_t bin_r_u08(struct Bin_R *r, BIN_ERR_READ *error);
> ```

I think it's more convenient to have the error argument and the return value switched around, the error being changed to bool:

```C
bool bin_r_u08(struct Bin_R *r, uint8_t *dest);
```

That way you can put one or multiple short-circuited writes/reads in a single if-condition to check if the entire write/read succeeded

```C
if (!bin_r_u08(foo, &struct.var1) ||
     !bin_r_u16(foo, &struct.var2)) {
    // reads failed!
}
```

Instead of the mouthful

```C
BIN_ERR_READ error;
struct.var1 = bin_r_u08(foo, &error);
if (error != BIN_ERR_READ_OK) {
    // first read failed
}
struct.var2 = bin_r_u16(foo, &error);
if (error != BIN_ERR_READ_OK) {
    // second read failed
}
```

you can't just check if the entire read/write has failed in short-circuit way without writing all this out.

Though with short circuiting would won't know which exact function call failed, but you might not need to know, and if you do need to know, you can rewrite it as in this `BIN_ERR_READ` example too.

```C
bool success = true;
success &= bin_r_u08(foo, &struct.var1);
if (!success) {
    // first read failed
}
success &= bin_r_u16(foo, &struct.var2);
if (!success) {
    // second read failed
}
```

or if you you don't need short circuiting things

```C
bool success = true;
success &= bin_r_u08(foo, &struct.var1);
success &= bin_r_u16(foo, &struct.var2);
if (!success) {
   ...
}
```

Instead of the mouthful

```C
bool success = true;
BIN_ERR_READ error;
struct.var1 = bin_r_u08(foo, &error);
success &= error == BIN_ERR_READ_OK;
struct.var2 = bin_r_u16(foo, &error);
success &= error == BIN_ERR_READ_OK;
if (!success) {
   ...
}
```

---

*[toxcore/bin_io.h, line 212 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQcZUztyy5ga6cN4n1:-KZQcZUztyy5ga6cN4n2:be2d35k) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L212)):*
> ```objc
> void bin_r_arr(struct Bin_R *r, uint8_t *data, size_t length, BIN_ERR_READ *error);
> 
> #endif /* BIN_IO_H */
> ```

Do you think a function returning current position of how far have we read into the buffer or returning the number of bytes left to read would see any use?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[toxcore/bin_io.h, line 79 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQbreQ96mp99VbbVzM:-KZQdX49RhOazZoO6WV6:bnrvdlk) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L79)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Add: This pointer becomes invalid when `bin_w_u*()` or `bin_w_arr()` functions are called, as they might reallocate the internal buffer at a different address.
</blockquote></details>

Needs to be added to the documentation, I mean.

---

*[toxcore/bin_io.h, line 145 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQdf-NO9WTareJTCUk:-KZQdf-NO9WTareJTCUl:b24e09e) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L145)):*
> ```objc
>  *
>  * The reader is the data retrieval counterpart of the writer specified above.
>  * It reads integers in network byte order as written by the writer.
> ```

That description can be interpreted as reader returning values in network byte order, meaning and you have to change the byte order to the host byte order yourself if you want that. But I believe reads does the conversion from network byte order to host byte order for us?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[toxcore/bin_io.h, line 55 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQWS-pDAP9FzBww9kY:-KZSI7lhZf34t7IjBWR0:b7lf0b) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L55)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Why do you have `struct FOO foo` function arguments if you typedef'ed the struct?
</blockquote></details>

Apidsl bug. This struct definition is the right thing to do here, but then below it should be using the typedef, not the struct tag. Apidsl needs a bunch of fixes, but first we'll make do and then make it pretty.

---

*[toxcore/bin_io.h, line 66 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQacQkHsY67gfQIiy2:-KZSKFJLhLmEEaa3Kcwy:brity68) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L66)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

How about having a constructor that doesn't require specification of the inital capacity?

Current constructor encourages to sum sizes of all things we are going to write, e.g. `bin_w_new(size1 + size2 + SOME_CONSTANT_SIZE*6 +1 +7 + size44 + sizeof(uint64_t), &error);`, which is quite ugly. If the writer automatically resizes internal array, this should really be optional. The default initial capacity could be the maximum toxcore packet size (there is a define for that) or something else if you think that's a bad choice.
</blockquote></details>

I agree. Removed the initial size.

---

*[toxcore/bin_io.h, line 79 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQbreQ96mp99VbbVzM:-KZSKe6O2YNrJlbrVSkR:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L79)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Needs to be added to the documentation, I mean.
</blockquote></details>

Done.

---

*[toxcore/bin_io.h, line 145 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQdf-NO9WTareJTCUk:-KZSL58LGpqS8Tmcl2O2:b-83330m) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L145)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

That description can be interpreted as reader returning values in network byte order, meaning and you have to change the byte order to the host byte order yourself if you want that. But I believe reads does the conversion from network byte order to host byte order for us?
</blockquote></details>

PTAL.

---

*[toxcore/bin_io.h, line 185 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQW_oIGfV_gzPspiFf:-KZSLqfRlQJ9_KQfpQT0:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L185)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

I think it's more convenient to have the error argument and the return value switched around, the error being changed to bool:

```C
bool bin_r_u08(struct Bin_R *r, uint8_t *dest);
```

That way you can put one or multiple short-circuited writes/reads in a single if-condition to check if the entire write/read succeeded

```C
if (!bin_r_u08(foo, &struct.var1) ||
     !bin_r_u16(foo, &struct.var2)) {
    // reads failed!
}
```

Instead of the mouthful

```C
BIN_ERR_READ error;
struct.var1 = bin_r_u08(foo, &error);
if (error != BIN_ERR_READ_OK) {
    // first read failed
}
struct.var2 = bin_r_u16(foo, &error);
if (error != BIN_ERR_READ_OK) {
    // second read failed
}
```

you can't just check if the entire read/write has failed in short-circuit way without writing all this out.

Though with short circuiting would won't know which exact function call failed, but you might not need to know, and if you do need to know, you can rewrite it as in this `BIN_ERR_READ` example too.

```C
bool success = true;
success &= bin_r_u08(foo, &struct.var1);
if (!success) {
    // first read failed
}
success &= bin_r_u16(foo, &struct.var2);
if (!success) {
    // second read failed
}
```

or if you you don't need short circuiting things

```C
bool success = true;
success &= bin_r_u08(foo, &struct.var1);
success &= bin_r_u16(foo, &struct.var2);
if (!success) {
   ...
}
```

Instead of the mouthful

```C
bool success = true;
BIN_ERR_READ error;
struct.var1 = bin_r_u08(foo, &error);
success &= error == BIN_ERR_READ_OK;
struct.var2 = bin_r_u16(foo, &error);
success &= error == BIN_ERR_READ_OK;
if (!success) {
   ...
}
```
</blockquote></details>

Done.

---

*[toxcore/bin_io.h, line 212 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQcZUztyy5ga6cN4n1:-KZSLrSlPDqSOBM8yB4I:b7r2v05) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L212)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Do you think a function returning current position of how far have we read into the buffer or returning the number of bytes left to read would see any use?
</blockquote></details>

Maybe, but YAGNI. I've also removed the "clone" function, because I don't know whether we're going to need it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[toxcore/bin_io.h, line 185 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQW_oIGfV_gzPspiFf:-KZSjfVu7cZTX5l9hh-N:bs6buo) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L185)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same for the writer.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
CMake builds fail because of missing `#include <stdbool.h>`. Autotools builds pass tough, since it''s not integrated in there.

iphydf commented:
Included stdbool.

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[toxcore/bin_io.h, line 185 at r1](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZQW_oIGfV_gzPspiFf:-KZSo-th6TNoVqt-msCR:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/30dc3974b74a3115e252e3e4f616da7a0eb5a8da/toxcore/bin_io.h#L185)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Same for the writer.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-:-KZSo1BP-SLswI-I5Rab:b-drgw2o)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks broke.

---

*[toxcore/bin_io.h, line 55 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZT6Ev4xzYpUosgdesy:-KZT6Ev5uPrTkjHeUQy1:btrlmz0) ([raw file](https://github.com/toktok/c-toxcore/blob/e175ac2a2e29ae3f2cddda73590a4ba114b9d209/toxcore/bin_io.h#L55)):*
> ```objc
>  * bin_w_size() will return 0 on the returned writer.
>  */
> struct Bin_W *bin_w_new(BIN_ERR_NEW *error);
> ```

What about constructors retrning null on the error?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[toxcore/bin_io.h, line 55 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZT6Ev4xzYpUosgdesy:-KZZaaSk0VEDL9lUDPdO:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/e175ac2a2e29ae3f2cddda73590a4ba114b9d209/toxcore/bin_io.h#L55)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

What about constructors retrning null on the error?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[toxcore/bin_io.h, line 145 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZ_sP2vvDtW8NKZjaMh:-KZ_sP2wZNRTLocuIwpi:beb9bc) ([raw file](https://github.com/toktok/c-toxcore/blob/d940cf36ee4fdc09645e5743d797760be45c9ba9/toxcore/bin_io.h#L145)):*
> ```objc
>  * Read an 8 bit unsigned integer and advance the reader by 1 byte.
>  */
> bool bin_r_u08(struct Bin_R *r, uint8_t *v);
> ```

Any reason for `u08` instead `u8` in the function name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
@iphydf How is this binary vector API? It's a bit complicated to have a PR based on a PR, that's why I just decided to push it in here. Say if you want me to create a separate PR for it based on your iphy/bin-io branch.

The function documentation might be worded funny, point out anything that needs to be fixed.

Also, how do I make apidsl prefix functions with 'bv_' instead of 'byte_vector_'?

GrayHatter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> clearly easier to write code with, but is this something we're really lacking from toxcore?

---

Reviewed 1 of 4 files at r1, 1 of 4 files at r2, 2 of 2 files at r4, 2 of 4 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 202 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7rEJPp84GiN0gMsI:-K_N7rEJPp84GiN0gMsJ:bcd7cd6) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/CMakeLists.txt#L202)):*
> ```CMake
> 
> # TODO(iphy, nurupo): move this in a separate library shared among toxcore, toxav, etc.
> # LAYER ?: ???
> ```

I'm not okay with using this in the crypto layer, so that leaves the network layer being the only one that makes sense.

---

*[toxcore/bin_io.h, line 28 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_NBZfDGmYr2Lzx7bxw:-K_NBZfDGmYr2Lzx7bxx:b-hj0em6) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/bin_io.h#L28)):*
> ```objc
>  * it returns false and puts the writer in the error state. Subsequent writes
>  * will have no effect and always return false. Currently, the only reason a
>  * writer function may fail is memory allocation failure.
> ```

Do we need a way to pre-allocate the size needed?

Do we need a function to return the error state?

---

*[toxcore/byte_vector.h, line 100 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N6NbBSvEgOyqjiSlx:-K_N6NbBSvEgOyqjiSly:bydpqel) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L100)):*
> ```objc
>  * Shrink the internal buffer to fit the contained data, making capacity
>  * be equal to the size. The internal buffer is reallocated if the vector if
>  * not already shrinked to fit.
> ```

Shrink the internal buffer to fit the contained data, setting capacity equal to the size. The internal buffer is reallocated if the vector if not already at the specified size.

---

*[toxcore/byte_vector.h, line 108 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N6znbRnn5lKqiIrwv:-K_N6znbRnn5lKqiIrww:b-h5o660) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L108)):*
> ```objc
>  * Extend vector at the specified position by inserting data starting there.
>  * Whatever is at and past the specified position gets relocated to new
>  * positions. The internal buffer is automatically reallocated if the current
> ```

`automatically resized`

---

*[toxcore/byte_vector.h, line 116 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7BNZtvDrv7CUcW7o:-K_N7BN_kv2H3XEPl9oi:bfjhsy0) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L116)):*
> ```objc
>  * @return True on success, false on failure.
>  */
> bool byte_vector_insert(struct Byte_Vector *_vector, size_t pos, const uint8_t *src, size_t count);
> ```

`s/count/size`

---

*[toxcore/byte_vector.h, line 135 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7QmwkWLhBvVBTBtl:-K_N7QmwkWLhBvVBTBtm:b-k5czct) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L135)):*
> ```objc
>  * @param count Number of bytes of the data to erase.
>  * @return True on success, false on failure.
>  */
> ```

What's the plan when parameters given don't match the available data.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-:-K_NCEOQoerqKprXphL-:b-b37fb3)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 202 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7rEJPp84GiN0gMsI:-K_NMfwoR_oVC0lbfcvd:b-va1ug8) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/CMakeLists.txt#L202)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

I'm not okay with using this in the crypto layer, so that leaves the network layer being the only one that makes sense.
</blockquote></details>

I think it should be a separate library, not a toxcore layer, as I can see it being used in several toxcore layers, as well as in toxav. That's what my comment above says.

If you look at changes, I changed `LAYER 1: Crypto core` to `LAYER ?: ???` because I don't think it belongs in any particular layer.

---

*[toxcore/bin_io.h, line 28 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_NBZfDGmYr2Lzx7bxw:-K_NLv-UiHqGbGHs4QLN:b-3jeuai) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/bin_io.h#L28)):*
>Do we need a way to pre-allocate the size needed?

No. If you read previous comments you will find that there was such a way and we decided to remove it.

---

*[toxcore/byte_vector.h, line 108 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N6znbRnn5lKqiIrwv:-K_NLWL37bgqmPiK1JDb:b-8olhc7) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L108)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

`automatically resized`
</blockquote></details>

No, `reallocated` is the desired wording. See the description of `$data()`.

---

*[toxcore/byte_vector.h, line 135 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7QmwkWLhBvVBTBtl:-K_NL3dqEYQLCPBGF-90:b92v3va) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L135)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

What's the plan when parameters given don't match the available data.
</blockquote></details>

Return false.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[toxcore/bin_io.h, line 28 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_NBZfDGmYr2Lzx7bxw:-K_Nb9c3HMATQapxnVpn:b-g7wy2q) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/bin_io.h#L28)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

>Do we need a way to pre-allocate the size needed?

No. If you read previous comments you will find that there was such a way and we decided to remove it.

</blockquote></details>

and checking error state?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

nurupo commented:
Moved the byte vector into a separate PR, as requested by @iphydf.

GrayHatter commented:
Reviewed 2 of 4 files at r7.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[CMakeLists.txt, line 202 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7rEJPp84GiN0gMsI:-K_yilq9d6SP4qg6K-oO:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/CMakeLists.txt#L202)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

I think it should be a separate library, not a toxcore layer, as I can see it being used in several toxcore layers, as well as in toxav. That's what my comment above says.

If you look at changes, I changed `LAYER 1: Crypto core` to `LAYER ?: ???` because I don't think it belongs in any particular layer.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

iphydf commented:
Review status: 4 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[toxcore/bin_io.h, line 145 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZ_sP2vvDtW8NKZjaMh:-K_ykW4_TVOe5lz6yN4P:b-t6pys1) ([raw file](https://github.com/toktok/c-toxcore/blob/d940cf36ee4fdc09645e5743d797760be45c9ba9/toxcore/bin_io.h#L145)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Any reason for `u08` instead `u8` in the function name?
</blockquote></details>

Just for pretty alignment.

---

*[toxcore/bin_io.h, line 28 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_NBZfDGmYr2Lzx7bxw:-K_ykKoOXE8IVjXb-Llb:b1fxdnq) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/bin_io.h#L28)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

and checking error state?
</blockquote></details>

There is only one error condition: memory allocation failure. All other parts of these functions are total. If that changes, we can easily change the API (it's private after all).

---

*[toxcore/byte_vector.h, line 100 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N6NbBSvEgOyqjiSlx:-K_ykaKYdP2eMu21WPRt:b-tl5w9j) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L100)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

Shrink the internal buffer to fit the contained data, setting capacity equal to the size. The internal buffer is reallocated if the vector if not already at the specified size.
</blockquote></details>

Moved to #379.

---

*[toxcore/byte_vector.h, line 108 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N6znbRnn5lKqiIrwv:-K_ykgrhlFp71ArkwalO:b-tl5w9j) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L108)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

No, `reallocated` is the desired wording. See the description of `$data()`.
</blockquote></details>

Moved to #379.

---

*[toxcore/byte_vector.h, line 116 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7BNZtvDrv7CUcW7o:-K_ykhPW5OJx6BahBZs2:b-tl5w9j) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L116)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

`s/count/size`
</blockquote></details>

Moved to #379.

---

*[toxcore/byte_vector.h, line 135 at r6](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-K_N7QmwkWLhBvVBTBtl:-K_ykhvDaIGGxv9R9_HD:b-tl5w9j) ([raw file](https://github.com/toktok/c-toxcore/blob/06bac1b18002f60d7280ecf0f000b55f0c96cfbe/toxcore/byte_vector.h#L135)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Return false.
</blockquote></details>

Moved to #379.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320)*
<!-- Sent from Reviewable.io -->

GrayHatter commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-:-KaVEGSIvxDazeg8Qsr1:bg7igkv)*
<!-- Sent from Reviewable.io -->

robinlinden commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r7, 8 of 8 files at r9.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[auto_tests/tut_test.h, line 4 at r9](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-Kg5XbkgXtooBtrg_tbt:-Kg5XbkgXtooBtrg_tbu:b-wjcnnk) ([raw file](https://github.com/toktok/c-toxcore/blob/75d622335f834fa1d59203c7dc5c19e0d1315a20/auto_tests/tut_test.h#L4)):*
> ```objc
> #include <tut/tut_reporter.hpp>
> 
> #define TEST_GROUP(NAME)		\
> ```

Is it intentional this this file uses tabs instead of spaces for alignment?

---

*[toxcore/bin_io.h, line 145 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-KZ_sP2vvDtW8NKZjaMh:-Kg5Y7vcgAsNKcyoS8a6:bsyfbzs) ([raw file](https://github.com/toktok/c-toxcore/blob/d940cf36ee4fdc09645e5743d797760be45c9ba9/toxcore/bin_io.h#L145)):*
<details><summary><i>Previously, iphydf wrote…</i></summary><blockquote>

Just for pretty alignment.
</blockquote></details>

Not a huge fan of this since 0n tends to mean n in base 8.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/320#-:-Kg5YZFKRLtB9RLpHSPb:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

robinlinden commented:
Why was this PR closed?

nurupo commented:
Closed in favor of https://github.com/TokTok/c-toxcore/pull/379.
